vulnerabilities/models.py

```diff
@@ -628,6 +631,153 @@ def get_absolute_url(self):
+            "namespace": package.namespace,
+            "type": package.type,
+            "qualifiers": package.qualifiers,
+            "subpath": package.subpath,
+        }
+
+        if fix:
+            filter_dict["packagerelatedvulnerability__fix"] = True
+
+        return Package.objects.filter(**filter_dict).distinct()
+
+    def sort_by_version(self, packages_to_sort):
+        """
+        Use the univers version_class to sort the incoming list of vulnerabilities.models.Package
+        objects.
+        """
+        univers_version = RANGE_CLASS_BY_SCHEMES[packages_to_sort[0].type].version_class
+        return sorted(
+            packages_to_sort,
+            key=lambda x: univers_version(x.version),
+        )
+
+    @property
+    def fixed_package_details(self):
+        """
+        For a given package, report the closest subsequent fixed by version for each of that
+        package's vulnerabilities as well as the closest non-vulnerable version and the most recent
+        non-vulnerable version.
+        """
+        fixed_packages = self.get_fixed_packages(package=self, fix=True)
+        qs = self.vulnerabilities.filter(packagerelatedvulnerability__fix=False)
+        all_sibling_packages = self.get_fixed_packages(package=self, fix=False)
+
+        non_vuln_sibs = []
+        for sib in all_sibling_packages:
+            if sib.is_vulnerable is False:
+                non_vuln_sibs.append(sib)
+
+        univers_version_class = RANGE_CLASS_BY_SCHEMES[self.type].version_class
+        later_non_vuln_sibs = []
+
+        for non_vuln_sib in non_vuln_sibs:
+            non_vuln_sib_univers_version = univers_version_class(
+                PackageURL.from_string(non_vuln_sib.purl).version
+            )
+
+            if univers_version_class(non_vuln_sib.version) > univers_version_class(self.version):
+                later_non_vuln_sibs.append(non_vuln_sib)
+
+        qs = qs.prefetch_related(
+            Prefetch(
+                "packages",
+                queryset=fixed_packages,
+                to_attr="fixed_packages",
+            )
+        )
+
+        purl_dict = {}
+        purl_dict["purl"] = PackageURL.from_string(self.purl)
+
+        closest_non_vulnerable_sib = ""
+        latest_non_vulnerable_sib = ""
+
+        if len(later_non_vuln_sibs) > 0:
+            closest_non_vulnerable_sib = self.sort_by_version(later_non_vuln_sibs)[0]
+            latest_non_vulnerable_sib = self.sort_by_version(later_non_vuln_sibs)[-1]
+
+            purl_dict["closest_non_vulnerable"] = PackageURL.from_string(
+                closest_non_vulnerable_sib.purl
+            )
+            purl_dict["latest_non_vulnerable"] = PackageURL.from_string(
+                latest_non_vulnerable_sib.purl
+            )
+
+        else:
+
+            purl_dict["closest_non_vulnerable"] = None
+            purl_dict["latest_non_vulnerable"] = None
+
+        purl_dict.update({"vulnerabilities": []})
+
+        for vuln in qs:
```

Comment:
The logic for the `for` loop should be in a different function, and should be tested separately

Reply:
@TG1999 I'm working on your last comment -- re moving the for loop logic to a new function.  I've done that (and it works) and am now creating the new test -- my new function requires 4 arguments, which I'm mocking in the test.  

FYI the new function `get_vulnerability_details()` is called like this:

```
        for vuln in package_vulnerabilities:
            package_details["vulnerabilities"].append({"vulnerability": vuln})
            self.get_vulnerability_details(vuln, version_class, fixed_by_packages, package_details)
```

That seems like a lot of data to keep passing to the function but as far as I can tell is necessary given the goal of the function (a nested vulnerability dictionary for each vulnerability of the package the user is searching for).

Reply:
@TG1999 Working today with @pombredanne, among other things I've extracted the for loop to a new function and created several other similar or helper functions/property methods.

In addition to fielding new comments/suggestions, I still need to
- update some existing tests and write a few new tests
- finish my work on the Package API and write 1 or more tests for that as well.

Reply:
This has been resolved.